src/crypto/chacha.cpp

```diff
@@ -0,0 +1,215 @@
+/*
+chacha-merged.c version 20080118
+D. J. Bernstein
+Public domain.
+*/
+
+#include <crypto/chacha.h>
+
+/* $OpenBSD: chacha.c,v 1.1 2013/11/21 00:45:44 djm Exp $ */
+
+typedef unsigned char u8;
+typedef unsigned int u32;
+
+typedef struct chacha_ctx chacha_ctx;
+
+#define U8C(v) (v##U)
+#define U32C(v) (v##U)
+
+#define U8V(v) ((u8)(v)&U8C(0xFF))
+#define U32V(v) ((u32)(v)&U32C(0xFFFFFFFF))
+
+#define ROTL32(v, n) (U32V((v) << (n)) | ((v) >> (32 - (n))))
+
+#define U8TO32_LITTLE(p)                                                       \
+  (((u32)((p)[0])) | ((u32)((p)[1]) << 8) | ((u32)((p)[2]) << 16) |            \
+   ((u32)((p)[3]) << 24))
+
+#define U32TO8_LITTLE(p, v)                                                    \
+  do {                                                                         \
+    (p)[0] = U8V((v));                                                         \
+    (p)[1] = U8V((v) >> 8);                                                    \
+    (p)[2] = U8V((v) >> 16);                                                   \
+    (p)[3] = U8V((v) >> 24);                                                   \
+  } while (0)
+
+#define ROTATE(v, c) (ROTL32(v, c))
+#define XOR(v, w) ((v) ^ (w))
+#define PLUS(v, w) (U32V((v) + (w)))
+#define PLUSONE(v) (PLUS((v), 1))
+
+#define QUARTERROUND(a, b, c, d)                                               \
+  a = PLUS(a, b);                                                              \
+  d = ROTATE(XOR(d, a), 16);                                                   \
+  c = PLUS(c, d);                                                              \
+  b = ROTATE(XOR(b, c), 12);                                                   \
+  a = PLUS(a, b);                                                              \
+  d = ROTATE(XOR(d, a), 8);                                                    \
+  c = PLUS(c, d);                                                              \
+  b = ROTATE(XOR(b, c), 7);
+
+static const char sigma[] = "expand 32-byte k";
+static const char tau[] = "expand 16-byte k";
+
+void chacha_keysetup(chacha_ctx *x, const u8 *k, u32 kbits) {
+  const char *constants;
+
+  x->input[4] = U8TO32_LITTLE(k + 0);
+  x->input[5] = U8TO32_LITTLE(k + 4);
+  x->input[6] = U8TO32_LITTLE(k + 8);
+  x->input[7] = U8TO32_LITTLE(k + 12);
+  if (kbits == 256) { /* recommended */
+    k += 16;
+    constants = sigma;
+  } else { /* kbits == 128 */
+    constants = tau;
+  }
+  x->input[8] = U8TO32_LITTLE(k + 0);
+  x->input[9] = U8TO32_LITTLE(k + 4);
+  x->input[10] = U8TO32_LITTLE(k + 8);
+  x->input[11] = U8TO32_LITTLE(k + 12);
+  x->input[0] = U8TO32_LITTLE(constants + 0);
+  x->input[1] = U8TO32_LITTLE(constants + 4);
+  x->input[2] = U8TO32_LITTLE(constants + 8);
+  x->input[3] = U8TO32_LITTLE(constants + 12);
+}
+
+void chacha_ivsetup(chacha_ctx *x, const u8 *iv, const u8 *counter) {
```

Comment:
I can confirm this function in 09ec49ba31867187cecd631593d37bcc5f3ca72f is identical to the original at de624c626ea081929df000b09932dbc804eda51d.